Infinite nils in ranges (#46)
andycroll/andycroll.com: _posts/2019/2019-05-12-watch-out-for-nils-in-ranges.md

---
title: Watch Out For nil in Ranges
description: To Infinity and Beyond
layout: article
category: ruby
image:
  base: '2019/watch-out-for-nil-in-ranges'
  alt: 'Buzz Lightyear'
---

Ruby 2.6 enhanced the syntax for ranges to allow for infinite values at either end. The new syntax is described in the [Ruby documentation for `Range`](http://ruby-doc.org/core-2.6.3/Range.html#class-Range-label-Endless+Ranges).

> This is great for representing concepts that require that, for example date ranges between a date and into the future, but the new “sugared” syntax changes existing behaviour.

Reviewer: ```suggestion
This is great for representing concepts such as date ranges between some date and one anytime the future, but this new “sugared” syntax changes existing behaviour.
```

## Instead of…

> …raising an `ArgumentError` when a `nil` is passed into a range…

Reviewer: I'm confused. I'm thinking: Why is `ArgumentError` being discussed here? I don't think the intro has set me to understand what this example is going to show.

```ruby
end_of_range = nil
(1..end_of_range).map { |i| do_something(i) }
#=> ArgumentError (bad value for range)
```

## Beware…

> …that a `Range` ending in `nil` no longer throws an error, but represents an _endless_ range.

Reviewer: I think you need to talk about this in the intro.

```ruby
end_of_range = nil
(1..end_of_range).map { |i| do_something(i) }
#=> infinite loop!
```

## But why?

Ruby 2.5, and earlier, already had the ability to create endless ranges, but only by using special constants—such as `Float::INFINITY` in the `Range`s.

> But allowing `nil` to represent, in a language where the result of methods can often _be_ `nil` definitely gaves us another place to make mistakes.

Reviewer: ```suggestion
By allowing `nil` to be considered a valid input to a range, in a language where the result of methods can often _be_ `nil` definitely gives us another place to make mistakes.
```

Reviewer: Or maybe something like: it's yet another potential trap...or can lead to confusing behaviour. Something like that.

## Why not?

> You don't really have a choice here, it’s a part of the language.

Reviewer: I think you need to set this whole newsletter up as: warning: a change of the language means you're likely to get caught out. Here's how to guard against it.

And then this 'why not?' section will make more sense.

Reviewer: And also: a full-stop or dash between these two clauses. Not a comma.

> But, you can protect yourself by ensuring that you check any value you pass into a `Range`. It isn’t “good practice” to scatter `nil` checks in your code, but Ruby has never been about purity in its approach to object-oriented programming.

Reviewer: This sentence reads a bit unclear...like I said, I think the whole narrative arc of this needs adjusting..then I'll be able to better comment on the closing.
